### DISCUSSION TITLE
Question about pos point of array.

### DISCUSSION THREAD

**Discussion by Dora138** [MaJerle/STM32_USART_DMA_RX#30] [Q&A]
Hi guys, 
This article is super fascinating. But I didn't understand on some of line. For example the line below:
pos = ARRAY_LEN(usart_rx_dma_buffer) - LL_DMA_GetDataLength(DMA1, LL_DMA_CHANNEL_5);

Why does The new point on array equal array length subduct the getDataLength?

If we received 5 characters and buffers length is 8, that means pos=3. 

Does anybody can help me to figure out what mistakes do I have.

**Comment by MaJerle** [comment]
Article at first page explains it. A new calculated position is the array index up to which we have valid new data in our rx buffer.

And old_pos stores value of last read, which now becomes new start read position.

**Reply by Dora138** [reply]
Hi Majerle,
I'm so appreciate on your quick response. And so sorry to bother you again.
After review carefully, I'm still puzzled. 

pos is the array index up to which we have valid new data in our rx buffer. That is clear.

ARRAY_LEN()is the length of array. That is clear.

LL_DMA_GetDataLength is we received data from DMA.

If we give a buffer 20, we received 5 from DMA.

As we know, we need read the data from pos0 to pos4
In the buffer. Then set the read point to pos4.

But the code confuses me pos=20-5 =15. 

It looks like read data from pos0 to pos15 when the old_pos=0 at beginning.

Please let me know how can I understand that question or some of parameters which I misunderstand.

**Comment by MaJerle** [comment]
Function get data length returns "remainig data length". So if we receive 5 bytes and initial size was 20, function will return 15. 20-15 = 5.

**Reply by Dora138** [reply]
You teach me. Thanks again.
